### PR TITLE
Create Cabal-syntax for .cabal parsing and types

### PR DESCRIPTION
**Pull request by patrickdoc**
---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

---

This PR splits `Cabal` into `Cabal-syntax` and `Cabal` (re: #7559).

The changes were generally conservative to avoid making design decisions. I mostly just pulled PackageDescription.hs and GenericPackageDescription.hs into the new package, and then moved code over as necessary to fix all the compilation errors.

The two code changes I did make were to move a few utility parsing functions so that Distribution.Verbosity stayed in `Cabal` rather than `Cabal-syntax`. Otherwise it was just import changes to avoid going through the Distribution.Simple.Utils module.

The result is split almost exactly in half (126 modules in `Cabal-syntax`, 121 modules in `Cabal`)

### CODE REVIEW

File: Cabal-syntax/Cabal-syntax.cabal
Comment (phadej) [inline Cabal-syntax/Cabal-syntax.cabal:141]: Most of these modules should be `exposed-modules`. In fact, the different lists could be combiled.

I don't think any module have to be hidden (i.e. `other-module`), except maybe some `Compat` modules, though I'm not sure about these either.

File: Cabal-syntax/Cabal-syntax.cabal
Comment (fgaz) [inline Cabal-syntax/Cabal-syntax.cabal:82]: We may want to split the compat stuff between Cabal and Cabal-syntax, like it's done now with cabal-install and Cabal, but I'm not sure it's worth doing in this case.

File: Cabal-syntax/src/Distribution/Fields/ParseResult.hs
Comment (phadej) [inline Cabal-syntax/src/Distribution/Fields/ParseResult.hs:170]: Moving these seems wrong. (probably related that this module have to exposed?)

**Review by phadej — COMMENTED** [review]
This indeed looks very simple.

Could you make changes I mentioned.

Making CI pass will need some changes there, but I can take a look at that.

Note: GHC devs will need to be aware of this change, their build system will need adjustments too.

File: Cabal-syntax/Cabal-syntax.cabal
Comment (phadej) [reply Cabal-syntax/Cabal-syntax.cabal:82]: I'd leave them for now, and just drop pre-GHC-8 support (#7531) before next major. That would cleanup plenty of stuff, leaving mostly "We rather not depend on `dlist` package and make our own copy".

**Comment by phadej** [comment]
Most (maybe even all?) `test-suites` in `Cabal` can be moved to `Cabal-syntax`. These might highlight if things are in wrong place too (or imported from wrong place):

```
HackageTests.hs:import Distribution.Simple.Utils                   (fromUTF8BS, toUTF8BS)
UnitTests.hs:import Distribution.Simple.Utils
UnitTests.hs:import qualified UnitTests.Distribution.Simple.Glob
UnitTests.hs:import qualified UnitTests.Distribution.Simple.Program.GHC
UnitTests.hs:import qualified UnitTests.Distribution.Simple.Program.Internal
UnitTests.hs:import qualified UnitTests.Distribution.Simple.Utils
UnitTests.hs:    , testGroup "Distribution.Simple.Glob"
UnitTests.hs:        UnitTests.Distribution.Simple.Glob.tests
UnitTests.hs:    , UnitTests.Distribution.Simple.Program.GHC.tests
UnitTests.hs:    , testGroup "Distribution.Simple.Program.Internal"
UnitTests.hs:        UnitTests.Distribution.Simple.Program.Internal.tests
UnitTests.hs:    , testGroup "Distribution.Simple.Utils" $
UnitTests.hs:        UnitTests.Distribution.Simple.Utils.tests ghcPath
UnitTests/Distribution/Compat/Time.hs:import Distribution.Simple.Utils (withTempDirectory)
UnitTests/Distribution/Simple/Glob.hs:module UnitTests.Distribution.Simple.Glob
UnitTests/Distribution/Simple/Glob.hs:import Distribution.Simple.Glob
UnitTests/Distribution/Simple/Program/GHC.hs:module UnitTests.Distribution.Simple.Program.GHC (tests) where
UnitTests/Distribution/Simple/Program/GHC.hs:import Distribution.Simple.Program.GHC (normaliseGhcArgs)
UnitTests/Distribution/Simple/Program/GHC.hs:tests = testGroup "Distribution.Simple.Program.GHC"
UnitTests/Distribution/Simple/Program/Internal.hs:module UnitTests.Distribution.Simple.Program.Internal
UnitTests/Distribution/Simple/Program/Internal.hs:import Distribution.Simple.Program.Internal ( stripExtractVersion )
UnitTests/Distribution/Simple/Utils.hs:module UnitTests.Distribution.Simple.Utils
UnitTests/Distribution/Simple/Utils.hs:import Distribution.Simple.BuildPaths ( exeExtension )
UnitTests/Distribution/Simple/Utils.hs:import Distribution.Simple.Utils
```

Looks like very few unit tests are testing `Distribution.Simple`.

**Comment by phadej** [comment]
`Distribution.Simple.Glob` could be renamed to `Distribution.Utils.Glob`.

File: Cabal-syntax/src/Distribution/Fields/ParseResult.hs
Comment (patrickdoc) [reply Cabal-syntax/src/Distribution/Fields/ParseResult.hs:170]: So I moved these because "Verbosity" seems like a cabal-install specific thing as opposed to a Cabal-syntax thing, so I left it in `Cabal`.

These seem to be mostly for nice error reporting in a context where "Verbosity" makes sense, but I would guess most external use cases would rather handle the errors differently.

File: Cabal-syntax/Cabal-syntax.cabal
Comment (patrickdoc) [reply Cabal-syntax/Cabal-syntax.cabal]: Updated

File: Cabal-syntax/src/Distribution/Fields/ParseResult.hs
Comment (phadej) [reply Cabal-syntax/src/Distribution/Fields/ParseResult.hs:170]: Good point about `Verbosity`.

**Comment by patrickdoc** [comment]
Running into some small compatibility issues, and not sure what the standard practices are here.

1) I've locally moved the .cabal specific checks of this module: https://github.com/haskell/cabal/blob/master/Cabal/src/Distribution/PackageDescription/Check.hs to `Cabal-syntax`, but one of the checks requires `autogenPathsModuleName` (to verify `Paths_Cabal`, etc.). This is defined in Distribution.Simple.BuildPaths, and it doesn't seem reasonable to me to move all of those build paths into `Cabal-syntax`. Not sure if `Paths_Cabal` is an official part of the `PackageDescription` syntax, or if it's just an implementation detail of `Cabal`.

2) Moving `Distribution.Package` to `Cabal-syntax` breaks `hackage-security`. https://github.com/haskell/hackage-security/search?p=1&q=Distribution It would be pretty trivial to move `hackage-security` to use `Cabal-syntax`, but not sure how these types of changes are usually handled. Otherwise I could add a compatibility layer, but there might be some weirdness with name conflicts.

**Comment by gbaz** [comment]
I'm not sure if moving the checks to cabal-syntax makes sense or not as a whole? I would thing the checks are for the _semantics_ of a well-formed cabal file, not for parsing and manipulating its syntax. I can see arguments either way, though.

On the latter, could you use reexported modules? https://cabal.readthedocs.io/en/3.4/cabal-package.html#pkg-field-library-reexported-modules

I'm not sure if that fits in the support window for Cabal features, but I think it might?

**Comment by phadej** [comment]
@gbaz, consider `hackage-server`. I think `Cabal-syntax` is enough for it, and `hackage-server` uses `PD.Check`. Also `hackage-cli` could use check. So i'd rather have that in `Cabal-syntax`, though I'm not against having `Cabal-check` lib too, but maybe that is too much at once!

**Comment by gbaz** [comment]
As the comments say:

```
-- This has code for checking for various problems in packages. There is one
-- set of checks that just looks at a 'PackageDescription' in isolation and
-- another set of checks that also looks at files in the package. Some of the
```

Hackage-server needs both types of checks, so I imagine it would need stuff that went beyond cabal-syntax regardless...

**Comment by phadej** [comment]
So? The package content checking interface is abstracted:

```haskell
-- | A record of operations needed to check the contents of packages.
-- Used by 'checkPackageContent'.
--
data CheckPackageContentOps m = CheckPackageContentOps {
    doesFileExist        :: FilePath -> m Bool,
    doesDirectoryExist   :: FilePath -> m Bool,
    getDirectoryContents :: FilePath -> m [FilePath],
    getFileContents      :: FilePath -> m BS.ByteString
  }

-- | Sanity check things that requires looking at files in the package.
-- This is a generalised version of 'checkPackageFiles' that can work in any
-- monad for which you can provide 'CheckPackageContentOps' operations.
--
-- The point of this extra generality is to allow doing checks in some virtual
-- file system, for example a tarball in memory.
--
checkPackageContent :: (Monad m, Applicative m)
                    => CheckPackageContentOps m
                    -> PackageDescription
                    -> m [PackageCheck]
```

---

Or put it other way around: `hackage-server` doesn't need to depend on `Distribution.Simple`, it doesn't build the packages. (builds do, but they probably use `cabal-install` or something higher level anyway!)

**Comment by gbaz** [comment]
Right, we _could_ put all those checks into cabal syntax. My understanding is those checks do not belong in cabal-syntax, because they're beyond the scope of cabal-syntax, which is, as the name says, _cabal syntax_, and checks for well-formedness of a tarball don't make sense there. I'm proposing we put no checks in cabal syntax, since most clients will not need them, or those that will, will want all of them, including the ones that don't belong there.

As one of the primary people currently maintaining hackage-server, I really don't care if it needs to continue to depend on the whole `Cabal` library -- that seems fine to me.

**Comment by patrickdoc** [comment]
Reexported modules were exactly what I needed, thanks.

Looking at the check code again, there wasn't really a compelling reason to move it in the first place. Just a little over-aggressive in trimming the test dependencies. I'll roll that back.

**Comment by patrickdoc** [comment]
Rolled back the PackageDescription Check change, updated the rest of the modules, and I think CI should work now.

**Comment by phadej** [comment]
The 

```haskell
-- | The name of the auto-generated Paths_* module associated with a package
autogenPathsModuleName :: PackageDescription -> ModuleName
autogenPathsModuleName pkg_descr =
  ModuleName.fromString $
    "Paths_" ++ map fixchar (prettyShow (packageName pkg_descr))
  where fixchar '-' = '_'
        fixchar c   = c
```

can very well be in `Cabal-syntax`. https://github.com/haskell/cabal/issues/6452 is an issue which would make the name very much relevant for `Cabal-syntax` part (though parsing doesn't change).

**Comment by phadej** [comment]
I tried this patch and it `cabal-fmt` and [`cabal-extras`](https://github.com/phadej/cabal-extras) work with this patch by changing `Cabal` to `Cabal-syntax`.

Except of `cabal-docspec` which is expected, as it does cpp preprocessing itself, i.e. very much "builds" the packages.

File: Cabal-QuickCheck/Cabal-QuickCheck.cabal
Comment (patrickdoc) [reply Cabal-QuickCheck/Cabal-QuickCheck.cabal:16]: `Cabal-QuickCheck` needs:

- `Distribution.Simple.Compiler  (DebugInfoLevel(..), OptimizationLevel(..), PackageDB(..),ProfDetailLevel(..),knownProfDetailLevels`
- `Distribution.Simple.Flag (Flag(..))`
- `Distribution.Simple.InstallDirs`
- `Distribution.Simple.Setup (HaddockTarget(..), TestShowDetails(..))`
- `Distribution.Utils.NubList`
- `Distribution.Verbosity`

`Cabal-described` needs `Distribution.Verbosity`
```
instance Described Verbosity where
    describe _ = REUnion
        [ REUnion ["0", "1", "2", "3"]
        , REUnion ["silent", "normal", "verbose", "debug", "deafening"]
          <> REMunch reEps (RESpaces <> "+" <>
            -- markoutput is left out on purpose
            REUnion ["callsite", "callstack", "nowrap", "timestamp", "stderr", "stdout" ])
        ]
```

**Comment by patrickdoc** [comment]
Updated artifact/bootstrap/meta CI, but not sure how to teach `doctest` about the new package.

**Comment by patrickdoc** [comment]
I think I've now fixed doctest and the Windows artifact issue, which just leaves the ghc 7.X failures due to reexported-modules not being supported.

**Comment by jneira** [comment]
> 
> 
> Reexported modules were exactly what I needed, thanks.
> 
> Looking at the check code again, there wasn't really a compelling reason to move it in the first place. Just a little over-aggressive in trimming the test dependencies. I'll roll that back.

minor (?) caveat, reexported modules dont work well with `stack repl` see https://github.com/commercialhaskell/stack/issues/5077#issuecomment-713511471
if they dont work with `stack repl` the project could not be loaded in hls neither for now

**Comment by phadej** [comment]
`reexported-modules` would also tightly tie `Cabal-syntax` and `Cabal` together: whenever there is a minor version of `Cabal-syntax`, there should be a minor version of `Cabal` too. That is probably not an issue for `Cabal`, but in general people have to be aware of that tradeoff. IMO `Cabal` don't need to re-export `Cabal-syntax`, though it would help in the migration, migration need is transient. (I expect very few projects having needs beyond `Cabal-syntax`).

**Comment by phadej** [comment]
That said. It would be very nice to check what `stack` uses from `Cabal`. I.e. how much extra it needs beyond `Cabal-syntax`.

From my code-only scan, it seems  it needs only macro generation (i.e. the same as `cabal-docspec`).

And now I'm starting to think where those belong. :(

**Comment by patrickdoc** [comment]
> `reexported-modules` would also tightly tie `Cabal-syntax` and `Cabal` together: whenever there is a minor version of `Cabal-syntax`, there should be a minor version of `Cabal` too.

To be clear, I was only reexporting the single module Distribution.Package so that hacakge-security didn't need to be patched for cabal-install to build successfully. I can also update hackage-security if that route is easier.

> That said. It would be very nice to check what `stack` uses from `Cabal`. I.e. how much extra it needs beyond `Cabal-syntax`.

From a quick look, the two most likely issues I saw were the `Setup.hs` files here: https://github.com/commercialhaskell/stack/search?q=Distribution.Simple

**Comment by phadej** [comment]
[`Distribution.Package`](https://hackage.haskell.org/package/Cabal-3.6.0.0/docs/src/Distribution.Package.html) re-exports other stuff. Still, my argument holds, even single (small) re-exported module ties the packages together.

**Comment by patrickdoc** [comment]
Any suggestions on how to deal with the breaking change for `hackage-security` without re-exporting `Distribution.Package`? It's needed for `cabal-install`, so we would need a patched version of `hackage-security` using the new `Cabal-syntax` for the builds to work.

**Comment by gbaz** [comment]
Just make a branch of hackage-security with the patch. Since its under the haskell org umbrella, we can make sure the patch is accepted when this is merged.

That said, I don't think reexported modules is problematic here. The concern about stack is about backpack, afaik, not reexported modules, which is a much earlier feature, and the version number coupling should not be a concern in our _particular_ case. (i.e. the idea of an update to the syntax lib without an update to the main Cabal lib seems unlikely)

**Comment by jneira** [comment]
>  The concern about stack is about backpack, afaik, not reexported modules, which is a much earlier feature

yeah, i overlooked it, i hit the issue using `mixins` and no `reexported-modules`

**Comment by patrickdoc** [comment]
Rebased, removed the reexport, and created a patch for hackage-security: https://github.com/haskell/hackage-security/pull/254

**Comment by fgaz** [comment]
@phadej do you think this is ready to be merged?

**Comment by patrickdoc** [comment]
@fgaz There's some extended conversation on the related hackage-security pr: https://github.com/haskell/hackage-security/pull/254#discussion_r744166675

It looks like the recommendation is to release a dummy Cabal-syntax 3.6 that helps create mutually exclusive bounds, then update/release hackage security with the linked pr, then merge this. I'm not sure how to set that in motion.

**Comment by Mikolaj** [comment]
I will help, I'm a maintainer of `hackage-security`. Tell me what to do.

**Review by Mikolaj — APPROVED** [review]
LGTM

**Comment by Mikolaj** [comment]
The hackage-security PR is merged, the dummy Cabal-syntax is on Hackage, all seems ready to go. Last moment to have a critical look before merge. This is quite a fabulous PR IMHO, but also very disruptive, so it'd be good not to revert and re-revert it too many times...

**Comment by phadej** [comment]
I think this is great overall (and we might or might not move some modules around in the future). However, to avoid reverts, you should make GHC devs aware that this is going to happen *soon*, cause if they realise this is too long, there will be too much pressure to revert indeed.

File: solver-benchmarks/solver-benchmarks.cabal
Comment (phadej) [inline solver-benchmarks/solver-benchmarks.cabal:49]: this package may depend on `Cabal & Cabal-syntax ==3.7.0.0` only, as its a test package

**Comment by Mikolaj** [comment]
> I think this is great overall (and we might or might not move some modules around in the future). However, to avoid reverts, you should make GHC devs aware that this is going to happen _soon_, cause if they realise this is too late (in GHC release pipeline), there will be too much pressure to revert indeed.

Than you for the heads up. I've just told the GHC hackers and will liaison with them as needed.

File: solver-benchmarks/solver-benchmarks.cabal
Comment (patrickdoc) [reply solver-benchmarks/solver-benchmarks.cabal]: Updated to only `Cabal-syntax >= 3.7`

File: Cabal/src/Distribution/Simple/PackageDescription.hs
Comment (fgaz) [inline Cabal/src/Distribution/Simple/PackageDescription.hs:12]: the `Module` haddock is inconsistent with the package name. Also why was this moved from `Cabal-syntax`'s `Distribution.PackageDescription.Parsec`?

**Review by fgaz — COMMENTED** [review]
Could you add `Cabal-syntax` in `README.md`, under "This Cabal Git repository contains the following packages:"?

File: Cabal/src/Distribution/Simple/PackageDescription.hs
Comment (fgaz) [reply Cabal/src/Distribution/Simple/PackageDescription.hs:12]: > Also why was this moved from Cabal-syntax's Distribution.PackageDescription.Parsec?

Ah I see, `Verbosity` is in cabal-install

**Comment by patrickdoc** [comment]
> Could you add `Cabal-syntax` in `README.md`, under "This Cabal Git repository contains the following packages:"?

Updated

File: Cabal/src/Distribution/Simple/PackageDescription.hs
Comment (patrickdoc) [reply Cabal/src/Distribution/Simple/PackageDescription.hs]: Yeah, these are mostly just error-handling wrappers around the actual parsers. And since they use `Verbosity` it felt better to keep them in `cabal-install` instead of letting that leak into `Cabal-syntax`.

**Comment by bgamari** [comment]
I suspect that it shouldn't be too hard to adapt GHC to this change but I'll have to try to know for certain.

**Comment by Mikolaj** [comment]
> I suspect that it shouldn't be too hard to adapt GHC to this change but I'll have to try to know for certain.

Should we wait?

File: Cabal-syntax/Cabal-syntax.cabal
Comment (Mikolaj) [reply Cabal-syntax/Cabal-syntax.cabal:44]: Indeed, we bumped this already on master, but mechanical rebasing did not reflect that in a PR a new package. I suppose the bounds need to be resynchronized with Cabal.cabal.

**Comment by bgamari** [comment]
> > I suspect that it shouldn't be too hard to adapt GHC to this change but I'll have to try to know for certain.
> 
> Should we wait?

I am currently working on attempting the adaptation. So far so good but I will feel better once both build systems have passed the testsuite. I'll let you know shortly.

**Comment by bgamari** [comment]
Is `Cabal-syntax/src/Distribution/Fields/Lexer.hs` guaranteed to be up-to-date?

**Comment by phadej** [comment]
> Is `Cabal-syntax/src/Distribution/Fields/Lexer.hs` guaranteed to be up-to-date?

Yes. `Cabal`s own CI doesn't use `alex` either. And there is a check that it should be up to date.

**Comment by bgamari** [comment]
I have confirmed that with the following patch I can bootstrap GHC:

```patch
diff --git a/Cabal-syntax/Cabal-syntax.cabal b/Cabal-syntax/Cabal-syntax.cabal
index 7649ac76b..91c04002d 100644
--- a/Cabal-syntax/Cabal-syntax.cabal
+++ b/Cabal-syntax/Cabal-syntax.cabal
@@ -41,7 +41,7 @@ library
     mtl        >= 2.1      && < 2.3,
     parsec     >= 3.1.13.0 && < 3.2,
     pretty     >= 1.1.1    && < 1.2,
-    text       >= 1.2.3.0  && < 1.3,
+    text       (>= 1.2.3.0  && < 1.3) || (>= 2.0 && < 2.1),
     time       >= 1.4.0.1  && < 1.12,
     -- transformers-0.4.0.0 doesn't have record syntax e.g. for Identity
     -- See also https://github.com/ekmett/transformers-compat/issues/35
diff --git a/Makefile b/Makefile
index 9af2ede22..2e218ef7f 100644
--- a/Makefile
+++ b/Makefile
@@ -23,15 +23,10 @@ lib-ghc-7.8 :
 
 # source generation: Lexer
 
-LEXER_HS:=Cabal-syntax/src/Distribution/Fields/Lexer.hs
+LEXER_HS:= Cabal-syntax/src/Distribution/Fields/Lexer.hs
 
 lexer : $(LEXER_HS)
 
-$(LEXER_HS) : templates/Lexer.x
-       alex --latin1 --ghc -o $@ $^
-       cat -s $@ > Lexer.tmp
-       mv Lexer.tmp $@
-
 # source generation: SPDX
 
 SPDX_LICENSE_HS:=Cabal-syntax/src/Distribution/SPDX/LicenseId.hs
diff --git a/ghc-packages b/ghc-packages
index 8d827f6f9..e9e369c42 100644
--- a/ghc-packages
+++ b/ghc-packages
@@ -1,2 +1,3 @@
 Cabal
+Cabal-syntax
```

**Comment by Mikolaj** [comment]
Well spotted re `/ghc-packages`.

**Comment by patrickdoc** [comment]
Fixed text and ghc-packages + bumped time to `< 1.13` to match `Cabal.cabal`

**Comment by Mikolaj** [comment]
I think we are go. :)

**Pull request merged**
